### PR TITLE
feat(IDX): enable bzlmod

### PR DESCRIPTION
**Pull request by nmattia**
This turns on [Bazel modules](https://bazel.build/external/module) and moves some of our (Bazel) dependency declarations to `MODULE.bazel`. This means those dependencies are specified declaratively and not imperatively as was the case in the `WORKSPACE`. This also allows us to update some dependencies, like `rules_python`, which previously created conflicts in transitive dependencies.

Some dependencies were also replaced (`rules_gazelle` -> `gazelle`).

This also adds a new workflow that ensures the Bazel lockfile, `MODULE.bazel.lock`, is up to date on all the platforms we support (Apple Intel, Apple Silicon, Linux x86). See workflow file for logic and details.

### CODE REVIEW

**Review by schneiderstefan — APPROVED** [review]

File: .github/workflows/bazel-lockfile.yml [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  # Merge groups should be checked for up-to-date lockfile
+  merge_group:
+  # Allows you to run this workflow manually from the Actions tab
+  workflow_dispatch:
+  push:
+    branches:
+      # master is checked for up-to-date lockfile
+      - master
+      # dev branches get a new commit with updated lockfile (when necessary)
+      - 'dev-gh-*'
```
Comment (basvandijk) [inline .github/workflows/bazel-lockfile.yml:17]: Shouldn't there be a `pull_request` trigger here?

Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:17]: good catch :)

File: .github/workflows/bazel-lockfile.yml [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+# runs for the same workflow are cancelled on PRs but not on master
+# (logic copied from main workflow)
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.ref || github.run_id }}
```
Comment (cgundy) [inline .github/workflows/bazel-lockfile.yml:22]: You can update this logic too, based on: https://github.com/dfinity/ic/pull/1958

File: .github/workflows/bazel-lockfile.yml [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+          BUILDEVENT_APIKEY: ${{ secrets.HONEYCOMB_API_TOKEN }}
+      - uses: actions/upload-artifact@v4
+        with:
+          name: bazel-module-lock-apple-intel
+          path: ./MODULE.bazel.lock
+
+  # Finally download the lockfile that went through every platform and check it
+  # for changes. See steps for actual logic depending on GHA event.
+  bazel-bzlmod-lockfile-update:
+    name: Check lockfile changes
```
Comment (cgundy) [inline .github/workflows/bazel-lockfile.yml:154]: If this is triggered on the merge_group, you will need to add this as a required job as well - otherwise it will be hanging forever.

File: requirements.txt [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #
-# This file is autogenerated by pip-compile with python 3.10
-# To update, run:
+# This file is autogenerated by pip-compile with Python 3.8
```
Comment (cgundy) [inline requirements.txt:2]: This seems weird.... why has it downgraded to python 3.8?

Comment (cgundy) [reply requirements.txt:2]: Did you manually run it from your env?

File: BUILD.bazel [hunk 5]
```diff
@@ -79,19 +79,19 @@ gazelle(
...
-    timeout = "moderate",
-    requirements_in = "requirements.in",
+    src = "requirements.in",
+    # NOTE: the requirements.txt reports the system python's version, not
+    # the one specified in MODULE.bazel.
+    # https://rules-python.readthedocs.io/en/latest/pypi-dependencies.html#pip-rules
     requirements_txt = "requirements.txt",
-    tags = [
-        "requires-network",
-    ],
```
Comment (cgundy) [inline BUILD.bazel:94]: Why did you remove this? I believe it does require the network as it needs to check the latest version.

File: .github/workflows/bazel-lockfile.yml (hunk 2 above)
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:22]: Thanks! will do

File: .github/workflows/bazel-lockfile.yml (hunk 3 above)
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:154]: Good point! will do; after merge obviously :)

File: BUILD.bazel (hunk 5 above)
Comment (nmattia) [reply BUILD.bazel:94]: It only needs network when you `bazel run` it AFAICT (`bazel run //:python-requirements.update`) . For regular builds it definitely shouldn't need network

File: requirements.txt (hunk 4 above)
Comment (nmattia) [reply requirements.txt:2]: See NOTE in the BUILD.bazel:

```
    # NOTE: the requirements.txt reports the system python's version, not
    # the one specified in MODULE.bazel.
    # https://rules-python.readthedocs.io/en/latest/pypi-dependencies.html#pip-rules
 ```

File: BUILD.bazel (hunk 5 above)
Comment (cgundy) [reply BUILD.bazel:94]: Okay now I'm not sure anymore... I know that `compile_pip_requirements` generates a test called `python-requirements_test` that validates if the requirements are still up to date and that failed in the past when `requires-network` was not specified. But it looks like it succeeded on your PR, so maybe that is not an issue anymore.

File: .github/workflows/bazel-lockfile.yml [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+          BRANCH_NAME: ${{ github.head_ref || github.ref_name }}
+          CI_RUN_ID: ${{ github.run_id }}
+          RUSTFLAGS: "--remap-path-prefix=${CI_PROJECT_DIR}=/ic"
+          BUILDEVENT_DATASET: "github-ci-dfinity"
+        with:
+          BAZEL_COMMAND: "build"
+          BAZEL_TARGETS: "//..."
+          BAZEL_EXTRA_ARGS: '--lockfile_mode=update --nobuild'
+          BAZEL_CI_CONFIG: "--config=ci --repository_cache=/cache/bazel"
+          # check if PR title contains release and set timeout filters accordingly
```
Comment (basvandijk) [inline .github/workflows/bazel-lockfile.yml:97]: Was this comment accidentally copied from somewhere else since it seems out of place here?

File: .github/workflows/bazel-lockfile.yml [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        with:
+          name: bazel-module-lock-apple-silicon
+          path: ./MODULE.bazel.lock
+
+  bazel-bzlmod-lockfile-linux:
+    name: Linux
+    container:
+      image: ghcr.io/dfinity/ic-build@sha256:115daa5ad5149182bb0416cbe5730f305be3bb2f48df576bc2c23067eefce84b
+      options: >-
+        -e NODE_NAME --privileged --cgroupns host -v /cache:/cache -v /var/sysimage:/var/sysimage -v /var/tmp:/var/tmp -v /ceph-s3-info:/ceph-s3-info
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:53]:
```suggestion
        -e NODE_NAME
```

File: .github/workflows/bazel-lockfile.yml [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        env:
+          CI_COMMIT_SHA: ${{ github.sha }}
+          CI_JOB_NAME: ${{ github.job }}
+          CI_JOB_URL: "${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"
+          CI_PIPELINE_SOURCE: ${{ github.event_name }}
+          CI_PROJECT_DIR: ${{ github.workspace }}
+          BRANCH_NAME: ${{ github.head_ref || github.ref_name }}
+          CI_RUN_ID: ${{ github.run_id }}
+          RUSTFLAGS: "--remap-path-prefix=${CI_PROJECT_DIR}=/ic"
+          BUILDEVENT_DATASET: "github-ci-dfinity"
```
Comment (basvandijk) [inline .github/workflows/bazel-lockfile.yml:91]: I think you want to disable [the diffing script](https://sourcegraph.com/github.com/dfinity/ic/-/blob/ci/bazel-scripts/main.sh?L38):
```suggestion
          RUN_ON_DIFF_ONLY: false
          BUILDEVENT_DATASET: "github-ci-dfinity"
```

File: .github/workflows/bazel-lockfile.yml [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+          path: ./MODULE.bazel.lock
+
+  bazel-bzlmod-lockfile-linux:
+    name: Linux
+    container:
+      image: ghcr.io/dfinity/ic-build@sha256:115daa5ad5149182bb0416cbe5730f305be3bb2f48df576bc2c23067eefce84b
+      options: >-
+        -e NODE_NAME --privileged --cgroupns host -v /cache:/cache -v /var/sysimage:/var/sysimage -v /var/tmp:/var/tmp -v /ceph-s3-info:/ceph-s3-info
+    runs-on:
+      group: zh1
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:55]: We don't need to be limited to `zh1` for this, do we?

**Review by cgundy — APPROVED** [review]

File: .github/workflows/bazel-lockfile.yml [hunk 10]
```diff
@@ -0,0 +1,221 @@
+# Workflow to automatically update the bazel lockfile on pull requests and dev
+# branches. In general, fails if lockfile is out of date.
+# For each platform we support, we run a dry (--nobuild) Bazel build, and then
+# upload the resulting lockfile as an artifact. The next platform downloads the
+# lockfile artifact, and then also runs a dry build, rinse, repeat.
+name: Update MODULE.bazel.lock
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:6]:
```suggestion
name: Update Bazel Lockfile
```

File: .github/workflows/bazel-lockfile.yml (hunk 9 above)
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:55]: nope, don't even need `dind-large` either. Just copied whatever we had in ci-main.yml.

File: .github/workflows/bazel-lockfile.yml [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+    needs: bazel-bzlmod-lockfile-apple-silicon
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Before script
+        id: before-script
+        shell: bash
+        run: |
+          [ -n "${NODE_NAME:-}" ] && echo "Node: $NODE_NAME"
+      - name: Login to Dockerhub
```
[the remark this answers is not in the document]
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:66]: I think images are still pulled even though we have `--nobuild`, though I mostly tried to keep consistent with the other files. Happy to remove it if you'd like?

File: .github/workflows/bazel-lockfile.yml (hunk 6 above)
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:97]: Like most things in this file I tried to (as much as possible) keep it consistent with ci-main.yml (basically copy pasted). Happy to remove though!

File: .github/workflows/bazel-lockfile.yml (hunk 8 above)
Comment (basvandijk) [reply .github/workflows/bazel-lockfile.yml:91]: Actually when thinking about it further, I don't think it hurts to have the diffing script be activated. So you can ignore my comment or not.

**Review by basvandijk — APPROVED** [review]
Nice! Just a heads-up that https://github.com/dfinity/ic/pull/1946 also modifies WORKSPACE.bazel so whoever merges last should make sure the right ubuntu_base is being used.

File: .github/workflows/bazel-lockfile.yml [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        with:
+          name: bazel-module-lock-apple-silicon
+          path: ./MODULE.bazel.lock
+
+  bazel-bzlmod-lockfile-linux:
+    name: Linux
+    container:
+      image: ghcr.io/dfinity/ic-build@sha256:115daa5ad5149182bb0416cbe5730f305be3bb2f48df576bc2c23067eefce84b
+      options: >-
+        -e NODE_NAME
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:53]: I see it should be `-e NODE_NAME -v /cache:/cache` since `--repository_cache=/cache/bazel` is used.

File: .github/workflows/bazel-lockfile.yml [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+          if [ "${{ github.event_name }}" == "pull_request" ]; then
+
+            echo "updating MODULE.bazel.lock"
+            git config user.email "infra+github-automation@dfinity.org"
+            git config user.name "IDX GitHub Automation"
+            git add MODULE.bazel.lock
+            git commit -m "Update MODULE.bazel.lock"
+            git push
+
+            exit 1
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:215]: Why `exit 1`?

File: .github/workflows/bazel-lockfile.yml [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+              [[ "${{ github.ref_name }}" =~ ^dev-gh-* ]]; then
+
+            echo "updating MODULE.bazel.lock"
+            git config user.email "infra+github-automation@dfinity.org"
+            git config user.name "IDX GitHub Automation"
+            git add MODULE.bazel.lock
+            git commit -m "Update MODULE.bazel.lock"
+            git push
+
+            exit 1
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:203]: Why `exit 1`?

**Review by mraszyk — APPROVED** [review]

File: .github/workflows/bazel-lockfile.yml [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+        with:
+          ref: ${{ github.head_ref || github.ref }}
+          # set the token in case we need to push the updated lockfile
+          token: ${{ secrets.IDX_PUSH_TO_PR }}
+      - uses: actions/download-artifact@v4
+        with:
+          name: bazel-module-lock-apple-intel
+      - name: check and push
+        run: |
+          echo git status
```
Comment (marko-k0) [inline .github/workflows/bazel-lockfile.yml:168]:
```suggestion
          set -e
          echo git status
```

**Review by oggy-dfin — APPROVED** [review]
The formal models related changes LGTM

File: .github/workflows/bazel-lockfile.yml (hunk 13 above)
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:215]: because the commit should be marked as red if changes are needed (the lockfile should always be up to date)

File: .github/workflows/bazel-lockfile.yml (hunk 15 above)
Comment (nmattia) [reply .github/workflows/bazel-lockfile.yml:168]: sure

**Review by aterga — APPROVED** [review]

File: .github/workflows/bazel-lockfile.yml (hunk 13 above)
Comment (marko-k0) [reply .github/workflows/bazel-lockfile.yml:216]: Not sure I understand the logic. CI job successfully completes updating the lock file when you bump some version, so CI job and entire workflow should / could be green. In most of the cases, if not in all, the workflow is interrupted and commit doesn't end up green.
